### DISCUSSION TITLE
Is it possible to support Wechat miniprogram?

### DISCUSSION THREAD

**Discussion by qucc** [jeeliz/jeelizFaceFilter#256] [Q&A]
I found comment about wechat in the file [jeelizFaceFilter.moduleNoDOM.js](https://github.com/jeeliz/jeelizFaceFilter/blob/master/dist/jeelizFaceFilter.moduleNoDOM.js)
![image](https://user-images.githubusercontent.com/1032477/142969613-9cade31b-3358-454f-8683-0b7e2c17e033.png)
Is there any demos or tutorial about this topic?

**Comment by xavierjs** [comment]
Yes it is possible.
`jeelizFaceFilter.moduleNoDOM.js` is made at this purpose.
The few DOM function are emulated in this module (since in WeChat miniprogram there are no DOM objects). So it can run out of the DOM.
Also the video is different with WeChat, it is not an HTML video element but an array of pixel colors. You need to init FaceFilter with:

```javascript
JEEFACEFILTERAPI.init({
   NNC: NEURALNETWORKMODEL,
   videoSettings: {
      videoElement: FAKEVIDEOELEMENT
    },
    // ... otherSettings
});
```
Where you init `FAKEVIDEOELEMENT` by:
```javascript
const arrayBuffer = new Uint8Array(vw*vh*4); // vw and vh are video width and height in pixels
  FAKEVIDEOELEMENT = {
    isFakeVideo: true, //always true
    arrayBuffer: arrayBuffer,// the WeChat video arrayBuffer
    videoHeight: vh, // height in pixels
    videoWidth:  vw, //width in pixels
    needsUpdate: true // boolean
  };
```

and `NEURALNETWORKMODEL` should be the parsed JSON content of the neural network model (not the URL or the string content)

Once you update `FAKEVIDEOELEMENT.arrayBuffer` by the content of a new video frame, you need to set  `FAKEVIDEOELEMENT.needsUpdate=true` so that facefilter knows that the frame has changed and can internally update the texture.

**Reply by wulixu** [reply]
Great!

**Comment by qucc** [comment]
Thanks for your reply , 
I don't know if i done it right.  Today i try the solution above and stuck with this errors.
**ERROR in ContextFF:  Cannot filter float textures**
![image](https://user-images.githubusercontent.com/1032477/143211625-c9a2ce7c-015e-41fa-8c0e-8564896d8895.png)

Here is demo
https://github.com/qucc/minicodear

**Reply by wulixu** [reply]
I think canvas has different methods or properties in HTML and WeChat, need to modify the dist files for fixing that.

another WeChat mini-program project integrated with tracking.js for the same purpose, please refer to https://github.com/sanyuered/WeChat-MiniProgram-WebAR

**Comment by xavierjs** [comment]
Hi,
Thank you for your feedback.
Indeed, even if WebGL is implemented for Wechat miniprograms, it is the version1 without extensions (or few of them). However we can make it. I have updated a new version of `jeelizFaceFilter.moduleNoDOM.js` which should fix the issue (the 3.0.5)

**Reply by qucc** [reply]
I  just test the 3.0.5 this morning , now it show this error. 
![image](https://user-images.githubusercontent.com/1032477/143370969-4be5fb85-2778-4ec0-a657-e3574a057f8e.png)

I also change these lines blow . but still get a error .
https://github.com/qucc/minicodear/blob/39f18559ec8feaacc990a6378c1ef976a3d818dc/index/index.js#L59
https://github.com/qucc/minicodear/blob/39f18559ec8feaacc990a6378c1ef976a3d818dc/libs/jeelizFaceFilter.moduleNoDOM.js#L67

![image](https://user-images.githubusercontent.com/1032477/143371636-c83b37cf-c704-4e13-9e4c-a3f2678197d7.png)

**Reply by xavierjs** [reply]
Thank you for your feedback. Indeed it seems that in WeChat Miniprogram the `createImageData` method is carried by the canvas itself instead of the context (https://developers.weixin.qq.com/miniprogram/en/dev/api/canvas/Canvas.createImageData.html) I will push a fix.

**Comment by xavierjs** [comment]
Hi,
I have pushed a potential fix (version 3.0.6). I hope it will help

**Reply by wulixu** [reply]
Thanks for your help，we want to share the first demo about wechat app to your demos if done!

**Reply by wulixu** [reply]
Just buy a beer for you, thanks again!

**Comment by wulixu** [comment]
Hi, xavierjs

I have tested the latest version you uploaded, then It can run normally in the  monitor of Weixin Devtool. unfortunately, It cannot be shown the cube within face in real phone, but the background video mesh is show normally. I did different experiments , It doesn't work. actually, I am not an professional in webgl area, ask for your help.
the latest code is here:  https://github.com/qucc/minicodear/tree/threecube
the office Three.js + WeApp tutorial is here : https://github.com/wechat-miniprogram/threejs-miniprogram/tree/master/example
![image](https://user-images.githubusercontent.com/3402267/143602053-ff94422d-3d3d-428b-aa3b-bce21626e873.png)

**Reply by wulixu** [reply]
By the way, the performance of rendering video in WeChat app is lower than H5 web page in the same physical phone. maybe need tune performance in future.

**Reply by wulixu** [reply]
When I used only myself THREE + Canvas drawing within face filter ready callback, got below exception.
code is here:
https://github.com/wechat-miniprogram/threejs-miniprogram/blob/master/example/test-cases/cubes.js
WebGL: INVALID_OPERATION: uniformMatrix4fv: location is not from current program.

**Reply by xavierjs** [reply]
I think that, in HTML5, the video element stays on the GPU when the WebGL texture with the video is updated.
In Wechat, since we get the video as an array, it had to go through the CPU, and then be streamed back to the GPU to update the WebGL texture. I think it is a important performance bottleneck. Unfortunately the only way to improve this would be to have the `<video>` element in WeChat too.

I cannot test WeChat miniprogram since I cannot open a WeChat account.

This issue seems related: https://stackoverflow.com/questions/54107987/wechat-webgl-minigame-not-working-on-a-device-but-works-in-devtool
There is an answer with 2 upvotes.

**Reply by wulixu** [reply]
thanks，I will try it.
for the performance issue, I want to use camera tag in wechat page for rendering video, in face filter canvas only draws face effects. any ideas？

**Reply by wulixu** [reply]
I have a test appid:wx7295c32613a7792a ,AppSecret: 0ab9f82b73cc8a7e7ccc58373f79c9fe
![image](https://user-images.githubusercontent.com/3402267/143662668-4985ddd7-7dda-4152-a9a0-10e764c9d7b0.png)
![image](https://user-images.githubusercontent.com/3402267/143662655-40b96205-f482-4694-bf27-56eaf8ce25bb.png)

**Reply by wulixu** [reply]
I tried stackoverflow solution, but it doesn't work. personally, it's the not real reason. because the office demo can work fine.
meanwhile, for the performance, so I decide to use camera tag for video rendering, canvas for face effect, JeelizFaceFilter is for basic data of face. so I hope JeelizFaceFilter's init method can accept null or undefine canvas, and can return the face pose data. current version, get error code (AN ERROR HAPPENS. ERROR CODE = NO_CANVASID).

**Reply by wulixu** [reply]
I tried to separate one alone canvas for facefilter module to get face data, use other canvas to redraw, but failed when to invoke init mehtod with fakedCanvas. I don't know why, cuz the details in your code is hidden. so the canvas argument may be optional to separate webgl staff from basic face data for more tech scenes.
![image](https://user-images.githubusercontent.com/3402267/143667899-b32fac4b-b187-42d6-a7d7-739e204ef716.png)
![image](https://user-images.githubusercontent.com/3402267/143667906-bee64a36-8264-41e1-857f-362149f44c6a.png)

**Reply by wulixu** [reply]
or use FackedCanvas with width,height or something for resizing in real canvas.

**Reply by xavierjs** [reply]
For the performance issue, displaying the `<video>` element instead of the FaceFilter canvas with the video rendering won't improve the performance I think.
Indeed, even if we display directly the `<video>` element instead of FaceFilter canvas, we still need to send the video content to a WebGL texture, internally, in Jeeliz FaceFilter, so that we can do face detection. This operation is slow since the video goes through a JavaScript array. Then once we have the video WebGL texture, displaying it on the canvas is really fast.
I display the FaceFilter canvas with the video texture instead of the original `<video>` element because it is possible to change the video color using a WebGL shader like in this tutorial: https://jeeliz.com/blog/tutorial-javascript-webgl-webcam-facial-filter-on-the-theme-of-matrix/

**Reply by xavierjs** [reply]
For the canvas you should init FaceFilter as you do with:
```javascript
JEELIZFACEFILTER.init({
  canvas: fakedCanvas
  //...
});
```

were `fakedCanvas` is initialized by:
```javascript
  // set the canvas dimension and wrap with fake DOM methods
  initCanvas(canvas) {
    let SystemInfo = wx.getSystemInfoSync()
    canvas.pixelRatio = SystemInfo.pixelRatio;
    canvas.width = SystemInfo.windowWidth * canvas.pixelRatio;
    canvas.height = SystemInfo.windowHeight * canvas.pixelRatio;
    
    canvas.addEventListener = function() {}
    canvas.removeEventListener = function () {}
  }
  initCanvas(fakedCanvas);
```

**Reply by wulixu** [reply]
when use fakedCanvas with your suggestion, I got this error: AN ERROR HAPPENS. ERROR CODE = GL_INCOMPATIBLE.
//fake canvas with object.
fakedCanvas = new Object();

**Reply by xavierjs** [reply]
The `fakedCanvas` main purpose is to create the WebGL context used by FaceFilter, so it should not be a generic object but the DOM canvas object.
I suggest to get it from the DOM in this way:
```javascript
wx.createSelectorQuery().select('#fakedWebgl').node(res => { 
  const fakedCanvas = res.node
  initCanvas(fakedCanvas)
```

I sent you an email (to wulixu.william@hotmail.com) in order to verify my WeChat account.

**Reply by wulixu** [reply]
I tried it, but failed! just like facefilter changes global webgl's configuration. because I just pass fakedcanvas to facefilter, and draw in real canvas, but it failed.

**Reply by wulixu** [reply]
If your request is ready, please let me know. I will go to bed after 0.5 hour.

**Reply by xavierjs** [reply]
I will send a new request for WeChat subscription tomorrow.
I have forked your github repository here: https://github.com/xavierjs/minicodear and I have changed the loading order and setup a PR here: https://github.com/qucc/minicodear/pulls. Can you test it? 

Also maybe it is an issue because THREE and FaceFilter are sharing the same context. Can you test the `noThree` branch of my fork: https://github.com/xavierjs/minicodear/tree/noThree
It should just display an alert box with "Face Detected" when the face is detected

**Reply by wulixu** [reply]
It can work, because we can see the detected data in log.

**Reply by wulixu** [reply]
and, your lib need to fix these errors. in my local version, I just remove theses.
![image](https://user-images.githubusercontent.com/3402267/143797150-d63058be-87b3-403b-8764-1ef606e2dd3c.png)

**Reply by xavierjs** [reply]
Hi @wulixu I have just pushed a fix to this bug (It seems that in WeChat the canvas has no `addEventListener` method). I hope it will help. The new version number is 3.0.7. I will have more time to work on WeChat issue and test it for real on Wednesday and Thursday this week.

**Reply by wulixu** [reply]
If you need any help for setting up mini-program env, pls contact with me anytime.

**Reply by xavierjs** [reply]
Hi @wulixu thank you for your help testing FaceFilter on Wechat is in my work pipeline, I just have some bugs to solve on another project before, I will come back to you once it is done. If you encounter new issues feel free to post them here.

**Reply by xavierjs** [reply]
Hi @wulixu I have sent you an email to your Github email ( wulixu.william@hotmail.com ) with my phone number to activate my WeChat registration.

**Comment by wulixu** [comment]
The first thing is to setup the env in your local machine. so you do some tests for yourself.
and you can refer some projects based tensorflow.js with facelandermark model in WeChat app,  here: https://github.com/sanyuered/WeChat-MiniProgram-AR-TFJS
It can be run fine. just like below.
tips: when you use WeChat dev tools, there are some few bugs when to debug mode, just go to preview your result with scan qrcode.

**Reply by xavierjs** [reply]
Hi @wulixu @qucc did you make it work with WeChat miniprograms? I would be glad to help but I need to get my WeChat account validated by a registered WeChat user (since I live in the US).

**Reply by wulixu** [reply]
Thanks for coming back. currently, we just use directly tensorflow.js, while JeelizFaceFilter needs to support miniprograms with webgl,  Three.js or something else. 
 
So, If you need any help, pls feel free to contact with via email wulixu.william@hotmail.com. or we can meet online at your local time 9:00~10:00AM today.

**Reply by wulixu** [reply]
If done, we can create an chat group on WeChat for the further help!

**Reply by xavierjs** [reply]
Hi @wulixu I have just sent you an email with my phone number to activate so that I can get my account validated on WeChat. I have redo the subscription process just now so it should not be expired.

**Reply by wulixu** [reply]
same error, if you are online now, we can try it again.

**Reply by xavierjs** [reply]
@wulixu I am online now, the confirmation page is opened right now, I don't touch my phone this time !

**Comment by xavierjs** [comment]
It should work on https://github.com/xavierjs/minicodear, `master` branch (it displays a 3D cube above user's head).
I have fixed a bug and I have updated the main FaceFilter script. I also use 2 separate canvases (1 for facefilter and video display, and 1 for THREE)

**Comment by xavierjs** [comment]
Hi @wulixu  Whats up on this issue?

**Reply by wulixu** [reply]
Sorry for late, It cannot work again.
I will push my co-workers to deep into your library working with WeChat mini-program. 
if any update, I will contact with you!
Thanks again!

**Reply by xavierjs** [reply]
Hi @wulixu 
Is it because of an update of Jeeliz FaceFilter?
Does https://github.com/xavierjs/minicodear work?
